docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+transform may want to produce more than one value for each input, or want to
+view all the data in a single collected view in order to make cross data
+decisions.
+
+``` csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // 1 => many (or none)
+    public static IncrementalValueSource<U> TransformMany<T, U>(this IncrementalValueSource<T> source, Func<T, IEnumerable<U>> func) => ...
+
+    // many => 1
+    public static IncrementalValueSource<U> BatchTransform<T, U>(this IncrementalValueSource<T> source, Func<IEnumerable<T>, U> func) => ...
+
+    // many => many (or none)
+    public static IncrementalValueSource<U> BatchTransformMany<T, U>(this IncrementalValueSource<T> source, Func<IEnumerable<T>, IEnumerable<U>> func) => ...
+}
+```
+
+In our above example we could use `BatchTransform` to collect the individual
+file paths collected, and convert them into a single collection:
+
+```csharp
+    // apply a 1-to-1 transform on each text, which represents extracting the path
+    var transformed = additionalTexts.Transform((text) => text.Path);
+
+    // batch the collected file paths into a single collection
+    var batched = transformed.BatchTransform(texts => texts);
+```
+
+The author could have equally combined these two steps into a single operation
+that utilizes LINQ:
+
+```csharp
+    // using System.Linq;
+    var singleOp = additionalTexts.BatchTransform(texts => texts.Select(text => text.Path));
+```
+
+**OPEN QUESTION** Should there be versions of
+`BatchTransform`/`BatchTransformMany` that take no transformation, and just
+perform the identity function as specified above?
+
+### Outputting values
+
+At some point in the pipeline the author will want to actually use the
+transformed data to produce an output, such as a `SourceText`. For this purpose
+there are 'terminating' extension methods that allow the author to provide the
+resulting data the generator produces. For instance there is a
+`GenerateSource(...)` method:
+
+``` csharp
+static partial class IncrementalValueSourceExtensions
+{
+    public internal static IncrementalGeneratorOutput GenerateSource<T>(this IncrementalValueSource<T> source, Action<SourceProductionContext, T> action) => ...
+}
+```
+
+That can be used by the author to supply `SourceText` to be appended to the
+compilation via the passed in `SourceProductionContext`:
+
+```csharp
+    // create an output that takes the file paths from the above batch and makes some user visible syntax
+    var output = batched.GenerateSource((sourceContext, filePaths) =>
+    {
+        sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}");
+    });
+
+```
+
+After creation, output nodes must be then registered with the pipeline context:
```

Comment:
is there any point in ever making an output that isn't registered?  if not, should we consider making this a single step?